src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BasePatternSwitchLocalRewriter.cs

```diff
@@ -122,11 +117,158 @@ protected virtual LabelSymbol GetDagNodeLabel(BoundDecisionDagNode dag)
                 return label;
             }
 
+            /// <summary>
+            /// A utility class that is used to scan a when clause to determine if it might assign a variable,
+            /// directly or indirectly. Used to determine if we can skip the allocation of pattern-matching
+            /// temporary variables and use user-declared variables instead, because we can conclude that they
+            /// are not mutated while the pattern-matching automaton is running.
+            /// </summary>
+            protected class WhenClauseMightAssignWalker : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+            {
+                private readonly bool _isSwitchStatement;
+                private bool _mightAssignSomething;
+
+                internal WhenClauseMightAssignWalker(bool isSwitchStatement)
+                {
+                    this._isSwitchStatement = isSwitchStatement;
+                }
+
+                public bool MightAssignSomething(BoundExpression expr)
+                {
+                    if (expr == null || expr.ConstantValue != null)
+                    {
+                        return false;
+                    }
+
+                    this._mightAssignSomething = false;
+                    this.Visit(expr);
+                    return this._mightAssignSomething;
+                }
+
+                public override BoundNode VisitReturnStatement(BoundReturnStatement node)
```

Comment:
>VisitReturnStatement [](start = 42, length = 20)

Are there tests for the additional visit methods in this iteration?

Reply:
Not each one individually, no. I will add a test comparing when a when clause might assign something vs not.

---
In reply to: [180574667](https://github.com/dotnet/roslyn/pull/25963#discussion_r180574667) [](ancestors = 180574667)